assets/javascripts/discourse/components/modal/chart-ui-builder.gjs

```diff
@@ -0,0 +1,162 @@
+import Component from "@glimmer/component";
+import { tracked } from "@glimmer/tracking";
+import { Input } from "@ember/component";
+import { fn } from "@ember/helper";
+import { on } from "@ember/modifier";
+import EmberObject, { action } from "@ember/object";
+import didInsert from "@ember/render-modifiers/modifiers/did-insert";
+import DButton from "discourse/components/d-button";
+import DModal from "discourse/components/d-modal";
+import i18n from "discourse-common/helpers/i18n";
+import { debounce } from "discourse-common/utils/decorators";
+
+export default class ChartUiBuilder extends Component {
+  @tracked config;
+  @tracked rows;
+  @tracked disabled = true;
+
+  constructor() {
+    super(...arguments);
+    this.initializeConfig();
+  }
+
+  @action
+  addRow() {
+    this.rows.push(this.initializeRow());
+    this.rows = this.rows;
+  }
+
+  @action
+  removeRow(rowToBeRemoved) {
+    this.rows = this.rows.filter((row) => row !== rowToBeRemoved);
+    if (this.rows.length === 0) {
+      this.rows = [this.initializeRow()];
+    }
+  }
+
+  @action
+  insertChart() {
+    this.args.model.toolbarEvent.addText(this.generateChartMarkup());
+    this.args.closeModal();
+  }
+
+  @debounce(100)
+  focusLastLabel(element) {
+    element.focus();
+  }
+
+  @debounce(100)
+  setDisabled() {
+    this.disabled = !this.rows.some((row) => this.isValidRow(row));
+  }
+
+  isValidRow(row) {
+    return row.label && row.value;
+  }
+
+  generateChartMarkup() {
+    const chartOptions = ["type='horizontalBar'"];
+    if (this.config.title) {
+      chartOptions.push(`title='${this.config.title}'`);
+    }
+    if (this.config.xAxisTitle) {
+      chartOptions.push(`xAxisTitle='${this.config.xAxisTitle}'`);
+    }
+    let markup = `[chart ${chartOptions.join(" ")}]`;
+    markup += "\n";
+    this.rows.forEach((row) => {
+      if (this.isValidRow(row)) {
+        markup += `${row.label} | ${row.value}`;
+        markup += "\n";
+      }
+    });
+    markup += "[/chart]";
+    return markup;
+  }
+
+  initializeConfig() {
+    this.config = EmberObject.create({
+      title: null,
+      xAxisTitle: null,
+    });
+    this.rows = [this.initializeRow()];
+  }
+
+  initializeRow() {
+    return EmberObject.create({
+      label: null,
+      value: null,
+    });
```

Comment:
Do those two need to be `EmberObject`s? Could `{}` be enough? (if not, would be good to use {} with `@tracked` props, or `TrackedObject`)